### PR TITLE
Use Eigen::Ref strides when mapping roboptim matrix with numpy matrix

### PR DESCRIPTION
**Pull request by jorisv**
Since Roboptim use Eigen::Ref the numpy wrapping of roboptim matrix is broken.

I have turn every PyArray_SimpleNewFromData into PyArray_NewFromDescr to use Eigen::Ref stride and fix the wrapping.

Also I fix some unit tests.

### CODE REVIEW

File: tests/to-python.stdout [hunk 1]
```diff
@@ -1,4 +1,4 @@
-[[ 0.  0.  0.]
- [ 0.  0.  0.]
- [ 0.  0.  0.]]
+[[0. 0. 0.]
+ [0. 0. 0.]
+ [0. 0. 0.]]
```
Comment (bchretien) [inline tests/to-python.stdout]: The error is due to a change in matrix printing in NumPy. I'm afraid this change will break the test for older NumPy versions. I was considering providing a custom formatter to force the formatting style no matter what the NumPy version is, but I don't know if that's gonna work.

File: tests/mpc.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        solver.setParameter("ipopt.output_file", "mpc.log")
+        solver.setParameter("ipopt.hessian_approximation", "limited-memory")
+        # We can check with the ipopt derivative checker that previously checked jacobian
+        # are not well forwarded to ipopt
+        solver.setParameter("ipopt.derivative_test", 'first-order')
+
+        solver.solve()
+        r = solver.minimum()
+
+        self.assertTrue(isinstance(r, roboptim.core.PyResult))
```
Comment (bchretien) [inline tests/mpc.py]: Does this problem have a global optimum that we could check?

File: tests/mpc.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
+#!/usr/bin/env python3
```
Comment (bchretien) [inline tests/mpc.py:1]: We should use `python` instead of `python3`.

**Comment by bchretien** [comment]
I'll try to test this locally ASAP (too bad my CI PR is not ready yet). Thanks! :+1:

File: tests/to-python.stdout (hunk 1 above)
Comment (bchretien) [reply tests/to-python.stdout]: See my attempt here: https://github.com/roboptim/roboptim-core-python/pull/8/commits/56ca346f9c21bc3c2198f73ba55dfb2d5d62b9ce

Comment (bchretien) [reply tests/to-python.stdout]: I merged my change in the other PR,

File: tests/mpc.py (hunk 2 above)
Comment (jorisv) [reply tests/mpc.py]: Since objective is quadratic and constraints linear a global optimum exist and is easy to find.

File: src/wrap.cc [hunk 4]
```diff
@@ -52,18 +52,25 @@ namespace roboptim
 
         npy_intp inputSize = static_cast<npy_intp> (this->inputSize ());
         npy_intp outputSize = static_cast<npy_intp> (this->outputSize ());
-
-        PyObject* resultNumpy =
-          PyArray_SimpleNewFromData (1, &outputSize, NPY_DOUBLE, result.data ());
+        npy_intp resultStride = static_cast<npy_intp>(
+            result.innerStride()*Eigen::Index(sizeof(result_ref::Scalar)));
```
Comment (bchretien) [inline src/wrap.cc:56]: `Eigen::Index` is not available in older Eigen versions (cf. [note of Eigen 3.3 release](http://eigen.tuxfamily.org/index.php?title=3.3#Index_typedef)), cf. the CI build:

```txt
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In member function `virtual void roboptim::core::python::Function::impl_compute(roboptim::GenericFunction<roboptim::EigenMatrixDense>::result_ref, roboptim::GenericFunction<roboptim::EigenMatrixDense>::const_argument_ref) const':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:56:34: error: `Index' is not a member of `Eigen'
             result.innerStride()*Eigen::Index(sizeof(result_ref::Scalar)));
                                  ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:58:36: error: `Index' is not a member of `Eigen'
             argument.innerStride()*Eigen::Index(sizeof(argument_ref::Scalar)));
                                    ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In member function `virtual void roboptim::core::python::DifferentiableFunction::impl_gradient(roboptim::core::python::DifferentiableFunction::gradient_ref, roboptim::core::python::DifferentiableFunction::const_argument_ref, roboptim::core::python::DifferentiableFunction::size_type) const':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:186:36: error: `Index' is not a member of `Eigen'
             gradient.innerStride()*Eigen::Index(sizeof(gradient_ref::Scalar)));
                                    ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:189:36: error: `Index' is not a member of `Eigen'
             argument.innerStride()*Eigen::Index(sizeof(argument_ref::Scalar)));
                                    ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In member function `virtual void roboptim::core::python::DifferentiableFunction::impl_jacobian(roboptim::core::python::DifferentiableFunction::jacobian_ref, roboptim::core::python::DifferentiableFunction::const_argument_ref) const':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:259:40: error: `Index' is not a member of `Eigen'
                 argument.innerStride()*Eigen::Index(sizeof(argument_ref::Scalar)));
                                        ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In member function `PyObject* detail::ParameterValueVisitor::operator()(roboptim::GenericFunction<roboptim::EigenMatrixDense>::const_vector_ref) const':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:842:27: error: `Index' is not a member of `Eigen'
           v.innerStride()*Eigen::Index(
                           ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In member function `PyObject* detail::StateParameterValueVisitor::operator()(roboptim::GenericFunction<roboptim::EigenMatrixDense>::const_vector_ref) const':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:934:27: error: `Index' is not a member of `Eigen'
           v.innerStride()*Eigen::Index(sizeof(roboptim::Function::vector_ref::Scalar)));
                           ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In function `PyObject* getSolverStateX(PyObject*, PyObject*)'':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:2468:30: error: `Index' is not a member of `Eigen'
     state->x().innerStride()*Eigen::Index(sizeof(solverState_t::argument_t::Scalar)));
                              ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc: In function `PyObject* toDict(T&) [with T = roboptim::Result; PyObject = _object]'':
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:2645:28: error: `Index' is not a member of `Eigen'
     result.x.innerStride()*Eigen::Index(sizeof(result_t::argument_t::Scalar)));
                            ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:2659:32: error: `Index' is not a member of `Eigen'
     result.value.innerStride()*Eigen::Index(sizeof(result_t::result_t::Scalar)));
                                ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:2673:38: error: `Index' is not a member of `Eigen'
     result.constraints.innerStride()*Eigen::Index(sizeof(result_t::result_t::Scalar)));
                                      ^
/home/travis/build/roboptim/roboptim-core-python/src/wrap.cc:2690:33: error: `Index' is not a member of `Eigen'
     result.lambda.innerStride()*Eigen::Index(sizeof(result_t::vector_t::Scalar)));
```

**Pull request closed without merging**